quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,1083 @@
+}
+
+impl PulsarSource {
+    pub async fn try_new(
+        ctx: Arc<SourceExecutionContext>,
+        params: PulsarSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let subscription_name = subscription_name(&ctx.index_id, &ctx.source_config.source_id);
+        info!(
+            index_id=%ctx.index_id,
+            source_id=%ctx.source_config.source_id,
+            topics=?params.topics,
+            subscription_name=%subscription_name,
+            "Create Pulsar source."
+        );
+
+        let pulsar = connect_pulsar(&params).await?;
+
+        let mut previous_positions = BTreeMap::new();
+        for topic in params.topics.iter() {
+            let partitions = pulsar.lookup_partitioned_topic(topic).await?;
+
+            for (partition, _) in partitions {
+                let partition_id = PartitionId::from(partition);
+                let position_opt = checkpoint.position_for_partition(&partition_id).cloned();
+
+                if let Some(position) = position_opt {
+                    previous_positions.insert(partition_id, position);
+                }
+            }
+        }
+
+        let pulsar = spawn_message_listener(
+            subscription_name.clone(),
+            params.clone(),
+            pulsar,
+            previous_positions.clone(),
+        )
+        .await?;
+
+        Ok(Self {
+            ctx,
+            pulsar,
+            params,
+            subscription_name,
+            previous_positions,
+            state: PulsarSourceState::default(),
+        })
+    }
+
+    fn process_message(
+        &mut self,
+        message: Message<PulsarMessage>,
+        batch: &mut BatchBuilder,
+    ) -> anyhow::Result<()> {
+        let current_position = msg_id_to_position(message.message_id());
+
+        let doc = match message.deserialize() {
+            Err(e) => {
+                warn!(error = ?e, "Failed to parse message from queue.");
+                self.state.num_invalid_messages += 1;
+                return Ok(());
+            }
+            Ok(doc) => doc,
+        };
+
+        self.add_doc_to_batch(&message.topic, current_position, doc, batch)
+    }
+
+    fn add_doc_to_batch(
+        &mut self,
+        topic: &str,
+        position: Position,
+        doc: String,
+        batch: &mut BatchBuilder,
+    ) -> anyhow::Result<()> {
+        if doc.is_empty() {
+            warn!("Message received from queue was empty.");
+            self.state.num_invalid_messages += 1;
+            return Ok(());
+        }
+
+        let partition = PartitionId::from(topic);
+        let num_bytes = doc.as_bytes().len();
+
+        if let Some(previous) = self.previous_positions.get(&partition) {
+            // We skip messages which are older than the previously recorded position.
+            // This is because Pulsar may replay messages which have not yet been acknowledged but
+            // are in the process of being indexed.
```

Comment:
> but are in the process of being indexed.

I got confused by this last part. Those messages are already indexed and published as the position is inferior to the previous one.

I would say `but have already been published. This can happen for example if an indexer crashes after a publish and before the ack sent to Pulsar.`